src/mlpack/methods/range_search/rs_model_impl.hpp

```diff
@@ -18,6 +18,216 @@
+        // Log::Info << "Building query tree..." << endl;
+        std::vector<size_t> oldFromNewQueries;
+        typename RSType::Tree queryTree(std::move(querySet), oldFromNewQueries,
+            leafSize);
+        // Log::Info << "Tree built." << endl;
+        // Timer::Stop("tree_building");
+
+        std::vector<std::vector<size_t>> neighborsOut;
+        std::vector<std::vector<double>> distancesOut;
+        rs->Search(&queryTree, range, neighborsOut, distancesOut);
+
+        // Remap the query points.
+        neighbors.resize(queryTree.Dataset().n_cols);
+        distances.resize(queryTree.Dataset().n_cols);
+        for (size_t i = 0; i < queryTree.Dataset().n_cols; ++i)
+        {
+          neighbors[oldFromNewQueries[i]] = neighborsOut[i];
+          distances[oldFromNewQueries[i]] = distancesOut[i];
+        }
+      }
+      else
+      {
+        // Search without building a second tree.
+        rs->Search(querySet, range, neighbors, distances);
+      }
+}
+
+//! Save parameters for Train.
+TrainVisitor::TrainVisitor(arma::mat&& referenceSet,
+                                       const size_t leafSize
+                                       ) :
+      referenceSet(std::move(referenceSet)),
+      leafSize(leafSize)
+{}
+
+//! Default Train on the given RSType instance.
+template<template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+void TrainVisitor::operator()(RSTypeT<TreeType>* rs) const
+{
+  if (rs)
+    return rs->Train(std::move(referenceSet));
+  throw std::runtime_error("no range search model initialized");
+}
+
+//! Train on the given RSType specialized for KDTrees.
+void TrainVisitor::operator()(RSTypeT<tree::KDTree>* rs) const
+{
+  if (rs)
+    return TrainLeaf(rs);
+  throw std::runtime_error("no range search model initialized");
+}
+
+//! Train on the given RSType specialized for BallTrees.
+void TrainVisitor::operator()(RSTypeT<tree::BallTree>* rs) const
+{
+  if (rs)
+    return TrainLeaf(rs);
+  throw std::runtime_error("no range search model initialized");
+}
+
+//! Train specialized for Octrees.
+void TrainVisitor::operator()(RSTypeT<tree::Octree>* rs) const
+{
+  if (rs)
+    return TrainLeaf(rs);
+  throw std::runtime_error("no range search model initialized");
+}
+
+//! Train on the given RSType considering the leafSize.
+template<typename RSType>
+void TrainVisitor::TrainLeaf(RSType* rs) const
+{
+   if (rs->Naive())
+      {
+        rs->Train(std::move(referenceSet));
+      }
+      else
+      {
+       std::vector<size_t> oldFromNewReferences;
+       typename RSType::Tree* tree =
+                    new typename RSType::Tree(std::move(referenceSet),
+                    oldFromNewReferences, leafSize);
+       rs->Train(tree);
+ 
+     // Give the model ownership of the tree and the mappings.
+     rs->treeOwner = true;
+     rs->oldFromNewReferences = std::move(oldFromNewReferences);
+  }
+      }
```

Comment:
oh thanks for pointing out...I'll fix the irregularities in coding style as soon as possible and also make sure not to repeat this thing in future.